server/src/delta_cache.rs

```diff
@@ -0,0 +1,286 @@
+use std::collections::VecDeque;
+use unleash_types::client_features::{DeltaEvent, ClientFeature, Segment};
+
+
+
+#[derive(Debug, Clone)]
+pub struct DeltaHydrationEvent {
+    event_id: i32,
+    features: Vec<ClientFeature>,
+    segments: Vec<Segment>,
+}
+
+
+#[derive(Debug, Clone)]
+pub struct DeltaCache {
+    max_length: usize,
+    events: VecDeque<DeltaEvent>,
+    hydration_event: DeltaHydrationEvent,
+}
+
+trait GetEventId {
+    fn get_event_id(&self) -> i32;
+}
+
+impl GetEventId for DeltaEvent {
+    fn get_event_id(&self) -> i32 {
+        match self {
+            DeltaEvent::FeatureUpdated { event_id, .. }
+            | DeltaEvent::FeatureRemoved { event_id, .. }
+            | DeltaEvent::SegmentUpdated { event_id, .. }
+            | DeltaEvent::SegmentRemoved { event_id, .. }
+            | DeltaEvent::Hydration { event_id, .. } => *event_id,
+        }
+    }
+}
+
+impl DeltaCache {
+    pub fn new(hydration_event: DeltaHydrationEvent, max_length: usize) -> Self {
+        let mut cache = DeltaCache {
+            max_length,
+            events: VecDeque::new(),
+            hydration_event: hydration_event.clone(),
+        };
+        cache.add_base_event_from_hydration(hydration_event);
+        cache
+    }
+
+    fn add_base_event_from_hydration(&mut self, hydration_event: DeltaHydrationEvent) {
+        if let Some(last_feature) = hydration_event.features.last().cloned() {
+            self.add_events(vec![DeltaEvent::FeatureUpdated {
+                event_id: hydration_event.event_id,
+                feature: last_feature,
+            }]);
+        }
+    }
+
+    pub fn add_events(&mut self, events: Vec<DeltaEvent>) {
+        for event in events.into_iter() {
+            self.events.push_back(event.clone());
+            self.update_hydration_event(event);
+            if self.events.len() > self.max_length {
+                self.events.pop_front();
+            }
+        }
+    }
+
+    pub fn get_events(&self) -> &VecDeque<DeltaEvent> {
+        &self.events
```

Comment:
Is the plan to kick this exact VecDeque over the HTTP boundary to downstream SDKs? I ask because if this needs to be massaged into something else first, it's going to mean doing a copy before we do a copy for sending it.

For large feature sets this can get really expensive

Reply:
99% of times we take 1-5 events out of the VecDeque and push that over HTTP.